### ISSUE TITLE
assemble() should validate that circuit has measure

### ISSUE THREAD

**Issue by ryan-woo** [Qiskit/qiskit-terra#2467]
### What is the expected enhancement?
The assembler should ensure that each circuit has a `measure` at the end before submission and raise an error if the circuit does not.

This would save some time where an invalid job would wait through a queue before it is determined to be invalid on the backend.

**Comment by taalexander** [comment]
This is also one more constraint that is related to #2354.

**Comment by jaygambetta** [comment]
I dont see the the problem. If you want to run a job without measurement you should be allowed. Not all qubits need to be measured. You in running it on a simulator may not want a measurement.

**Comment by ryan-woo** [comment]
Yes this is related to #2354. After some discussion, a warning should be given in backend.run() when running against a real backend.

**Comment by ajavadia** [comment]
Yes a warning by the backend is useful since many people forget to insert measurement. For example the qasm simulator issues a warning for non-measured circuits, while the statevector simulator does not.
https://github.com/Qiskit/qiskit-terra/blob/cd7cafaf9d6bb337d48855eeccd776c812fc2300/qiskit/providers/basicaer/qasm_simulator.py#L642

But I agree it should not fail, and it should not be qiskit that issues the warning, since the qobj may be intended for submission to a statevector/unitary simulator.

**Comment by ajavadia** [comment]
I'm moving this to the IBMQ provider.
